File: wizards/com/sun/star/wizards/document/Shape.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */
package com.sun.star.wizards.document;

import com.sun.star.awt.Point;
import com.sun.star.awt.Size;
import com.sun.star.beans.PropertyVetoException;
import com.sun.star.drawing.XControlShape;
import com.sun.star.drawing.XShape;
import com.sun.star.drawing.XShapes;
import com.sun.star.lang.XMultiServiceFactory;
import com.sun.star.lang.XServiceInfo;
import com.sun.star.text.TextContentAnchorType;
import com.sun.star.uno.Exception;
import com.sun.star.uno.UnoRuntime;
import com.sun.star.wizards.common.Helper;

/**
 * To change the template for this generated type comment go to
 * Window&gt;Preferences&gt;Java&gt;Code Generation&gt;Code and Comments
 */
public class Shape
{

    public XShape xShape;
    protected FormHandler oFormHandler;
    public XServiceInfo xServiceInfo;
    protected Point aPoint;
    protected Size aSize;
    protected XControlShape xControlShape;
    public XMultiServiceFactory xMSF;
    public XShapes xShapes;

    public Shape(FormHandler _oFormHandler, Point _aPoint, Size _aSize)
    {
        this.aPoint = _aPoint;
        this.aSize = _aSize;
        this.oFormHandler = _oFormHandler;
        createShape("com.sun.star.drawing.ControlShape");
    }

    public Shape(FormHandler _oFormHandler, String _sServiceName, Point _aPoint, Size _aSize)
    {
        try
        {
            this.aPoint = _aPoint;
            this.aSize = _aSize;
            this.oFormHandler = _oFormHandler;
            Object oShape = oFormHandler.xMSF.createInstance(_sServiceName);
            xShapes = UnoRuntime.queryInterface(XShapes.class, oShape);
            xServiceInfo = UnoRuntime.queryInterface(XServiceInfo.class, oShape);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    public Shape()
    {
    }

    private void createShape(String sServiceName)
    {
        try
        {
            xMSF = oFormHandler.xMSFDoc;
            Object oShape = xMSF.createInstance(sServiceName);
            xShape = UnoRuntime.queryInterface(XShape.class, oShape);
            xShape.setPosition(aPoint);
            if (aSize != null)
            {
                xShape.setSize(aSize);
            }
            else
            {
                xShape.setSize(new Size(1000, 100));
            }
            Helper.setUnoPropertyValue(xShape, "AnchorType", TextContentAnchorType.AT_PARAGRAPH);
            xServiceInfo = UnoRuntime.queryInterface(XServiceInfo.class, xShape);
            xControlShape = UnoRuntime.queryInterface(XControlShape.class, xShape);

        }
        catch (Exception e)
        {
            e.printStackTrace(System.err);
        }
    }

    public Size getSize()
    {
        return xShape.getSize();
    }

    public void setSize(Size _aSize)
    {
        try
        {
            xShape.setSize(_aSize);
        }
        catch (PropertyVetoException e)
        {
            e.printStackTrace(System.err);
        }
    }

    public Point getPosition()
    {
        return xShape.getPosition();
    }

    public void setPosition(Point _aPoint)
    {
        xShape.setPosition(_aPoint);
    }
}

File: test/source/helper/shape.cxx
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 100 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <test/helper/shape.hxx>
#include <sal/types.h>

#include <com/sun/star/drawing/XShape.hpp>
#include <com/sun/star/lang/XComponent.hpp>
#include <com/sun/star/lang/XMultiServiceFactory.hpp>
#include <com/sun/star/beans/PropertyVetoException.hpp>

#include <com/sun/star/uno/Reference.hxx>

using namespace css;

namespace apitest::helper::shape
{
uno::Reference<drawing::XShape>
    OOO_DLLPUBLIC_TEST createEllipse(const uno::Reference<lang::XComponent>& r_xComponent,
                                     const sal_Int32 nX, const sal_Int32 nY, const sal_Int32 nWidth,
                                     const sal_Int32 nHeight)
{
    return createShape(r_xComponent, u"Ellipse", nX, nY, nWidth, nHeight);
}

uno::Reference<drawing::XShape>
    OOO_DLLPUBLIC_TEST createLine(const uno::Reference<lang::XComponent>& r_xComponent,
                                  const sal_Int32 nX, const sal_Int32 nY, const sal_Int32 nWidth,
                                  const sal_Int32 nHeight)
{
    return createShape(r_xComponent, u"Line", nX, nY, nWidth, nHeight);
}

uno::Reference<drawing::XShape>
    OOO_DLLPUBLIC_TEST createRectangle(const uno::Reference<lang::XComponent>& r_xComponent,
                                       const sal_Int32 nX, const sal_Int32 nY,
                                       const sal_Int32 nWidth, const sal_Int32 nHeight)
{
    return createShape(r_xComponent, u"Rectangle", nX, nY, nWidth, nHeight);
}

uno::Reference<drawing::XShape> OOO_DLLPUBLIC_TEST
createShape(const uno::Reference<lang::XComponent>& r_xComponent, std::u16string_view r_aKind,
            const sal_Int32 nX, const sal_Int32 nY, const sal_Int32 nWidth, const sal_Int32 nHeight)
{
    uno::Reference<lang::XMultiServiceFactory> xMSF(r_xComponent, uno::UNO_QUERY_THROW);
    uno::Reference<drawing::XShape> xShape(
        xMSF->createInstance(OUString::Concat("com.sun.star.drawing.") + r_aKind + "Shape"),
        uno::UNO_QUERY_THROW);

    try
    {
        xShape->setPosition(awt::Point(nX, nY));
        xShape->setSize(awt::Size(nWidth, nHeight));
    }
    catch (const beans::PropertyVetoException&)
    {
    }

    return xShape;
}

} // namespace apitest::helper::shape

/* vim:set shiftwidth=4 softtabstop=4 expandtab cinoptions=b1,g0,N-s cinkeys+=0=break: */
